### ISSUE TITLE
How can I extract the boundary of the mesh?

### ISSUE THREAD

**Issue by SHAYUANPENG** [dengwirda/mesh2d#9]
Excellent work! I found this program very helpful. I'm wondering how can I extract the boundary nodes/edges of the triangular mesh？I'm using the mesh to set up a finite element model

**Comment by krober10nd** [comment]
Hello SHAYUANPENG, 

Darren and I have discussed this before and I took his advice before and implemented some algorithms based on many of his codes here in the mesh2d package.

The function extdom_edges2.m extracts the boundary edges and vertices (in no particular order). 

https://github.com/CHLNDDEV/OceanMesh2D/blob/master/utilities/extdom_edges2.m

You can then pass these boundary edges and some other tria arguments to list the boundary segments in a walking order. The function that accomplishes this is extdom_polygon.m 

https://github.com/CHLNDDEV/OceanMesh2D/blob/master/utilities/extdom_polygon.m


Hope this helps,

**Comment by dengwirda** [comment]
@SHAYUANPENG The `tricon2` routine can be used to compute various triangle- and edge-centred connectivity information. Boundary edges are those connected to only a single triangle, and boundary nodes are the ones indexed by boundary edges. 

Something like the following is hopefully the kind of thing you're looking to do: 

`% have a mesh [pp,tt]`
`[ee,tt] = tricon2(tt);`
`ebnd = ee(:,4) <= +0 ; % true for boundary edges`
`nbnd = false(size(pp,1),1);`
`nbnd(ee(ebnd,1:2)) = true ; % true for boundary nodes`

**Comment by SHAYUANPENG** [comment]
Thanks to krober10nd  and  dengwirda, thank you for your advices! I have tried the codes provided by dengwirda, and it worked. Here blow is my understanding of the code:

% have a mesh [pp,tt]    %pp represents the nodes list, and tt represents the triangle list
[ee,tt] = tricon2(tt);        %every row of ee stores a  edge represented by two indexed nodes  and two  indexed triangles sharing the edge, every row of tt stores the indexed nodes and edges of the triangle.
ebnd = ee(:,4) <= +0 ;   % which means the edge is owned by only one triangle.
nbnd = false(size(pp,1),1); 
nbnd(ee(ebnd,1:2)) = true ;   %check the boundary nodes

There must exist some zeros in the 4th column of ee, the tricon2 does this from the following code:

    ne = size(ee,1);         % ee :edge list,ne*2;
    ee = [ee, zeros(ne*1,3)] ; % extend ee
    ep = +3 * ones (ne*1,1)  ;% ne*1
    for ti = +1 : nt          % loop over every triangle
        ei = tt(ti,4) ;       % index of the frist edge of the triangle
        ee(ei,ep(ei)) = ti;   % this is the same to ee(ei,3) = ti, the frist triangle which owns the edge
        ej = tt(ti,5) ;       % index of the second edge of the triangle
        ee(ej,ep(ej)) = ti;   % this is the same to ee(ej,3) = ti, the frist triangle which owns the edge
        ek = tt(ti,6) ;       % index of the third edge of the triangle
        ee(ek,ep(ek)) = ti;   % this is the same to ee(ek,3) = ti, the frist triangle which owns the edge
    
        ep(ei) = ep(ei)+1 ;   %the 3th column of ee is occupied,turn to the 4th column of ee
        ep(ej) = ep(ej)+1 ;   %If the 4th column = 0, that row corresponds to a boundary edge
        ep(ek) = ep(ek)+1 ;   
    end
    Thanks !!!!!!!

**Comment by dengwirda** [comment]
@SHAYUANPENG Great - your outline above is exactly correct. Let me know if you have any other questions.

**Comment by SHAYUANPENG** [comment]
I don't know if this question is appropriate. Anyway, I generated a 3-D mesh by sweeping the 2-D triangular mesh along a curve.  The output data profile is *.cfx5, all the elements are triangular prisms. In the cfx5 profile, the boundary face are represented by [m,n], where m represents the index of the element which owns the face,  and n varies from 1-5, represents the five different faces of the  triangular prism respectively. The problem is that I can't get the correct boundary faces. I have done two things to define the boundary face:
1. Frist I thought the problem may stem from the order of nodes in the triangle, so I adjusted the order to make sure that the normal vectors of all the triangles in the 2-D mesh are along the same direction.
 vert(:,3) = 0;                                 %nodes list
for iTriangle=1:size(tria,1)
       point1 = tria(iTriangle,1);
       point1 = vert(point1,:);
       point2 = tria(iTriangle,2);
       point2 = vert(point2,:);
       point3 = tria(iTriangle,3);
       point3 = vert(point3,:);
       local_Sf = 0.5*cross(point2-point1,point3-point1); % normal vector
       if local_Sf(1,3)<0
       tria(:,[1 2])=tria(:,[2 1]) ;                                            %adjust the order
       end
end
vert = vert(:,1:2);
2. extract the boundary edges form the 2-D mesh, assign n for each edge.
[ee,tt] = tricon2(tria(:,1:3));
ebnd = ee(:,4) <= +0 ; % true for boundary edges
ebndIndex = find(ebnd == 1);      
ebndTria = ee(ebndIndex,3); % check the owner of the edge      

%check the boundary edge correspond to frist/2nd/3rd edge of the owner
% the frist layer of 3-D mesh, element index comes from ebndTria,
% the nth layer of 3-D mesh, elements indexes are ebndTria+(n-1)*size(tria,1)
BD = [ebndTria,zeros(size(ebndTria,1),1)];

itt1 = tt(ebndTria,4);
[is,pos]=ismember(ebndIndex,itt1);
index = find(is==1);
BD(index,2)=4;               %assign n
itt2 = tt(ebndTria,5);
[is,pos]=ismember(ebndIndex,itt2);
index = find(is==1);
BD(index,2)=3;               %assign n
itt3 = tt(ebndTria,6);
[is,pos]=ismember(ebndIndex,itt3);
index = find(is==1);
BD(index,2)=1;               %assign n
Thank you
